allennlp/common/registrable.py

```diff
@@ -0,0 +1,55 @@
+from collections import defaultdict
+from typing import TypeVar, Type, Dict, List  # pylint: disable=unused-import
+
+from allennlp.common.checks import ConfigurationError
+
+T = TypeVar('T')
+
+class Registrable:
+    """
+    Any class that inherits from ``Registrable`` gains access to a named registry for its
+    subclasses. To register them, just decorate them with the classmethod
+    ``@BaseClass.register(name)``.
+
+    After which you can call ``BaseClass.list_available()`` to get the keys for the
+    registered subclasses, and ``BaseClass.by_name(name)`` to get the corresponding subclass.
+    Note that the registry stores the subclasses themselves; not class instances.
+    In most cases you would then call ``from_params(params)`` on the returned subclass.
+
+    You can specify a default by setting ``BaseClass.default_implementation``.
+    If it is set, it will be the first element of ``list_available()``.
+    """
+
+    _registry = defaultdict(dict)  # type: Dict[type, Dict[str, type]]
```

Comment:
for some reason I was convinced that it wouldn't typecheck with `Type`, but it seems to, so I'll change it